marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/flow.sjs

```diff
@@ -219,14 +221,58 @@ class Flow {
     }
     let stepDefTypeLowerCase = (stepDetails.type) ? stepDetails.type.toLowerCase(): stepDetails.type;
     let stepName = stepRef.name || stepRef.stepDefinitionName;
+    let prov = this.datahub.prov;
     for (let content of this.writeQueue) {
+      let previousUris = this.datahub.hubUtils.normalizeToArray(content.previousUri || content.uri);
       let info = {
-        derivedFrom: content.previousUri || content.uri,
+        derivedFrom: previousUris,
         influencedBy: stepName,
         status: (stepDefTypeLowerCase === 'ingestion') ? 'created' : 'updated',
         metadata: {}
       };
-      let provResult = this.datahub.prov.createStepRecord(jobId, flowName, stepName, stepRef.stepDefinitionName, stepDefTypeLowerCase, content.uri, info);
+      let provResult;
+      if (prov.granularityLevel() === prov.FINE_LEVEL && content.provenance) {
```

Comment:
This is a meaty chunk of code that seems like it would benefit from being in an e.g. "buildFineProvenanceData" function (or "create" or "save" if it's persisting the data as well and not just building an in-memory data structure).

Reply:
@ryanjdew Since the function is inserting a bunch of documents, I would use "create" or "save" here instead of "build".

Reply:
It doesn't technically get saved until the prov.commit() call later on.